crates/kornia-dnn/src/rtdetr.rs

```diff
@@ -0,0 +1,174 @@
+        self.num_threads = num_threads;
+        self
+    }
+
+    /// Builds and returns an `RTDETRDetector` instance.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the `RTDETRDetector` if successful, or a `DnnError` if an error occurred.
+    pub fn build(self) -> Result<RTDETRDetector, DnnError> {
+        RTDETRDetector::new(self.model_path, self.ort_dylib_path, self.num_threads)
+    }
+}
+
+/// RT-DETR object detector.
+///
+/// This struct represents an instance of the RT-DETR object detection model.
+pub struct RTDETRDetector {
+    session: Session,
+}
+
+impl RTDETRDetector {
+    // TODO: default to hf hub
+    /// Creates a new `RTDETRDetector` instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `model_path` - Path to the RT-DETR model file.
+    /// * `ort_dylib_path` - Path to the ONNX Runtime dynamic library.
+    /// * `num_threads` - Number of threads to use for inference.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the `RTDETRDetector` if successful, or a `DnnError` if an error occurred.
+    pub fn new(
+        model_path: PathBuf,
+        ort_dylib_path: PathBuf,
+        num_threads: usize,
+    ) -> Result<Self, DnnError> {
+        // set the ort dylib path
+        std::env::set_var("ORT_DYLIB_PATH", ort_dylib_path);
+
+        // create the ort session
+        let session = Session::builder()?
+            .with_optimization_level(GraphOptimizationLevel::Level3)?
+            .with_intra_threads(num_threads)?
+            .commit_from_file(model_path)?;
+
+        Ok(Self { session })
+    }
+
+    /// Runs object detection on the given image.
+    ///
+    /// # Arguments
+    ///
+    /// * `image` - The input image as an `Image<u8, 3>`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing a vector of `Detection` objects if successful, or a `DnnError` if an error occurred.
+    pub fn run(&self, image: &Image<u8, 3>) -> Result<Vec<Detection>, DnnError> {
+        // TODO: explore pre-allocating memory for the image
+        // cast and scale the image to f32
+        let mut image_hwc_f32 = Image::from_size_val(image.size(), 0.0f32)?;
+        kornia_image::ops::cast_and_scale(image, &mut image_hwc_f32, 1.0 / 255.)?;
+
+        // convert to HWC -> CHW
+        let image_chw = image_hwc_f32.permute_axes([2, 0, 1]).as_contiguous();
+
+        // TODO: create a Tensor::insert_axis in kornia-rs
+        let image_nchw = Tensor::from_shape_vec(
+            [
+                1,
+                image_chw.shape[0],
+                image_chw.shape[1],
+                image_chw.shape[2],
+            ],
+            image_chw.into_vec(),
+            CpuAllocator,
+        )?;
+
+        // make the ort tensor
+        let ort_tensor = ort::Tensor::from_array((image_nchw.shape, image_nchw.into_vec()))?;
+
+        // run the model
+        let outputs = self.session.run(ort::inputs!["input" => ort_tensor]?)?;
```

Comment:
@decahedron1 how could we pre-allocate the output tensor here ?

Reply:
This seems like the perfect use case for [`OutputSelector`](https://docs.rs/ort/2.0.0-rc.4/ort/struct.OutputSelector.html)!